lua/nvim-treesitter/refactor/navigation.lua

```diff
@@ -33,24 +33,76 @@ function M.list_definitions(bufnr)
 
   local qf_list = {}
 
-  for _, def in ipairs(definitions) do
-    locals.recurse_local_nodes(def, function(_, node, _, match)
-      local lnum, col, _ = node:start()
-
-      table.insert(qf_list, {
-        bufnr = bufnr,
-        lnum = lnum + 1,
-        col = col + 1,
-        text = ts_utils.get_node_text(node)[1] or "",
-        kind = match and match:sub(1, 1) or ""
-      })
-    end)
+  for _, node in ipairs(definitions) do
+    local lnum, col, _ = node.node:start()
+    local type = string.upper(node.type:sub(1, 1))
+    local text = ts_utils.get_node_text(node.node)[1] or ""
+    table.insert(qf_list, {
+      bufnr = bufnr,
+      lnum = lnum + 1,
+      col = col + 1,
+      text = text,
+      type = type,
+    })
   end
 
   vim.fn.setqflist(qf_list, 'r')
   api.nvim_command('copen')
 end
 
+function M.list_definitions_toc(bufnr)
+  local winnr = api.nvim_get_current_win()
+  local bufnr = api.nvim_win_get_buf(winnr)
+  local definitions = locals.get_definitions(bufnr)
+
+  if #definitions < 1 then return end
+
+  local loc_list = {}
+
+  -- Force some types to act like they are parents
+  -- instead of neighbors of the next nodes.
+  local containers = {
+    ['function'] = true,
+    ['type'] = true,
+    ['method'] = true,
+  }
```

Comment:
This may depend on the grammar and capture groups, but it works fine for python and rst, and works ok for lua (this due you can have identifiers like `M.my_function`). We can revisit later if we need something "smarter" or more "clues" from the capture groups.